modules/openapi-generator/src/main/resources/python/python-experimental/model_utils.mustache

```diff
@@ -65,31 +65,53 @@ class OpenApiModel(object):
+            return None
+
         visited_composed_classes = kwargs.get('_visited_composed_classes', ())
         if (
             cls.discriminator is None or
             cls in visited_composed_classes
         ):
-            # we don't have a discriminator
-            # or we have already visited this class before and are sure that we
-            # want to instantiate it this time
+            # There is no discriminator.
+            # Or we have already visited this class before and are sure that we
+            # want to instantiate it this time. For example, this happens when
+            # we deserialize Dog from Animal. Dog contains an Animal instance. We go from Animal to Dog back to Animal
             return super(OpenApiModel, cls).__new__(cls)
 
-        oneof_anyof_classes = []
-        oneof_anyof_classes.extend(cls._composed_schemas.get('oneOf', ()))
-        oneof_anyof_classes.extend(cls._composed_schemas.get('anyOf', ()))
-        new_cls = cls.get_discriminator_class(kwargs)
+        # Get the name and value of the discriminator property.
+        # The discriminator name is obtained from the discriminator meta-data
+        # and the discriminator value is obtained from the input data.
+        discr_propertyname_py = list(cls.discriminator.keys())[0]
+        discr_propertyname_js = cls.attribute_map[discr_propertyname_py]
+        if discr_propertyname_js in kwargs:
+            discr_value = kwargs[discr_propertyname_js]
+        elif discr_propertyname_py in kwargs:
+            discr_value = kwargs[discr_propertyname_py]
+        else:
+            # The input data does not contain the discriminator property.
+            path_to_item = kwargs.get('_path_to_item', ())
+            raise ApiValueError(
+                "Cannot deserialize input data due to missing discriminator. "
+                "The discriminator property '%s' is missing at path: "
+                "'%s'" % (discr_propertyname_js, path_to_item)
+            )
+
+        new_cls = get_discriminator_class(cls, visited_composed_classes, discr_value)
         if new_cls is None:
-            disc_prop_name_py = list(cls.discriminator.keys())[0]
-            disc_prop_name_js = cls.attribute_map[disc_prop_name_py]
             path_to_item = kwargs.get('_path_to_item', ())
             disc_prop_value = kwargs.get(
-                disc_prop_name_js, kwargs.get(disc_prop_name_py))
+                discr_propertyname_js, kwargs.get(discr_propertyname_py))
             raise ApiValueError(
                 "Cannot deserialize input data due to invalid discriminator "
                 "value. The OpenAPI document has no mapping for discriminator "
                 "property '%s'='%s' at path: %s" %
-                (disc_prop_name_js, disc_prop_value, path_to_item)
+                (discr_propertyname_js, disc_prop_value, path_to_item)
             )
 
         if new_cls in visited_composed_classes:
```

Comment:
# if we are making an instance of a composed schema Descendent
            # which allOf includes Ancestor, then Ancestor contains
            # a discriminator that includes Descendent.
            # So if we make an instance of Descendent, we have to make an
            # instance of Ancestor to hold the allOf properties.
            # This code detects that use case and makes the instance of Descendant
            # For example:
            # When making an instance of Dog, _visited_composed_classes = (Dog,)
            # then we make an instance of Animal to include in dog._composed_instances
            # so when we are here, cls is Animal
            # cls.discriminator != None
            # cls not in _visited_composed_classes
            # new_cls = Dog
            # but we know we know that we already have Dog
            # because it is in visited_composed_classes
            # so make Animal here

Reply:
I have copied this as is. Please submit a pull request if you want to make changes.